plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/DockerContainerNameGenerator.java

```diff
@@ -10,32 +10,64 @@
  *******************************************************************************/
 package org.eclipse.che.plugin.docker.machine;
 
+import com.google.inject.Inject;
+
 import org.eclipse.che.api.core.model.machine.Machine;
 import org.eclipse.che.api.core.model.machine.MachineConfig;
 import org.eclipse.che.api.environment.server.ContainerNameGenerator;
+import org.eclipse.che.commons.annotation.Nullable;
 
+import javax.inject.Named;
+import javax.inject.Singleton;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import static java.lang.String.format;
+
 /**
  * This class is used for generation docker container name or parsing important information from docker container name.
  *
  * @author Alexander Andrienko
+ * @author James Drummond
+ * @author Alexander Garagatyi
  */
+@Singleton
 public class DockerContainerNameGenerator implements ContainerNameGenerator {
     private static final String NODE_HOST_GROUP    = "(/|(/[0-9a-z.-]+/))?";
     private static final String WORKSPACE_ID_GROUP = "(?<workspaceId>workspace[0-9a-z]+)";
     private static final String MACHINE_ID_GROUP   = "(?<machineId>machine[0-9a-z]+)";
+    private static final String SERVER_ID_GROUP   = "(?<serverId>server[0-9a-z]*)";
+
+    private static final String  CONTAINER_NAME_REGEX   = "^" +
+                                                          NODE_HOST_GROUP +
+                                                          WORKSPACE_ID_GROUP + "_" +
+                                                          MACHINE_ID_GROUP + "_" +
+                                                          "%s" + // che server ID group
+                                                          ".+$";
 
-    private static final String  CONTAINER_NAME_REGEX   = "^" + NODE_HOST_GROUP + WORKSPACE_ID_GROUP + "_" + MACHINE_ID_GROUP + "_.+$";
-    private static final Pattern CONTAINER_NAME_PATTERN = Pattern.compile(CONTAINER_NAME_REGEX);
+    private final Pattern containerNamePattern;
+    private final String  cheServerIdGroup;
+
+    @Inject(optional = true)
```

Comment:
Injection of this property is not declared in the PR. So this PR in current state doesn't fix mentioned issue.